### PR TITLE
HPCC-21463 Refactor CThorContiguousRowBuffer

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
NOT ready for review.  Pushed for smoketest results.

**Comment by ghalliday** [comment]
@jakesmith please can you review this.  I think it is complete, but I am hesitant about the scope for breaking things.  For instance there was a problem in roxie which was caught by only one case - on a 4 way data file.

File: rtl/eclrtl/rtlcommon.hpp [hunk 1]
```diff
@@ -7,15 +7,72 @@
...
-class ECLRTL_API CThorContiguousRowBuffer : implements IRowPrefetcherSource
+class ECLRTL_API CThorContiguousRowBuffer : public CContiguousRowBuffer, implements IRowPrefetcherSource
 {
 public:
-    CThorContiguousRowBuffer() {};
+    CThorContiguousRowBuffer() = default;
     CThorContiguousRowBuffer(ISerialStream * _in);
 
-    inline void setStream(ISerialStream *_in) { in = _in; maxOffset = 0; readOffset = 0; }
+    inline void setStream(ISerialStream *_in) { CContiguousRowBuffer::setStream(_in); }
```
Comment (jakesmith) [inline rtl/eclrtl/rtlcommon.hpp]: why override?

**Comment by jakesmith** [comment]
@ghalliday - I've looked through it a couple of times and can't spot any issues, and it looks clearer/cleaner.

What was the roxie issue caught ?

**Comment by ghalliday** [comment]
The problem in roxie was that it was calling inputSteam->tell() instead of deserizlizerSource.tell().  That caused problems because it no longer skips the input stream at the start of each row.  I also had a similar problem with calling inputStream->eos() which I worked around.  I plan to restructure the roxie code further anyway.

File: rtl/eclrtl/rtlcommon.hpp (hunk 1 above)
Comment (ghalliday) [reply rtl/eclrtl/rtlcommon.hpp]: I'll check.  It should possibly be clearing readOffset as well.

Comment (ghalliday) [reply rtl/eclrtl/rtlcommon.hpp]: Fixed that problem.

**Comment by ghalliday** [comment]
@richardkchapman please can you review this.  I think it will be better to merge before starting the next stage.

**Pull request merged**